docs/typescript/testing.md

```diff
@@ -2,92 +2,246 @@
-    });
-
-    runPromise = worker.run();
-    const connection = new Connection();
-    client = new WorkflowClient(connection.service);
-  });
+import { MockActivityEnvironment } from '@temporalio/testing';
+import { Context } from '@temporalio/activity';
+
+const env = new MockActivityEnvironment({ attempt: 2 });
+const result = await env.run(
+  async (x) => x + Context.current().info.attempt,
+  2
+);
+assert.equal(result, 4);
+```
 
-  after(async function () {
-    worker.shutdown();
-    await runPromise;
-  });
+### Heartbeats and cancellation
 
-  afterEach(() => {
-    axios.get.restore && axios.get.restore();
-  });
+`MockActivityEnvironment` is an [`EventEmitter`](https://nodejs.org/api/events.html#class-eventemitter) which emits a `heartbeat` event which you can use to listen on heartbeats emitted by the Activity.
 
-  it('returns correct result', async function () {
-    const result = await client.execute(example, { taskQueue: 'testhttp' });
-    assert.equal(result, 'The answer is 42');
-  });
+It also exposes a `cancel` method which cancels the Activity Context.
+
+```ts
+import { MockActivityEnvironment } from '@temporalio/testing';
+import { CancelledFailure, Context } from '@temporalio/activity';
+
+const env = new MockActivityEnvironment();
+
+env.on('heartbeat', (d: unknown) => {
+  if (d === 6) {
+    env.cancel('test');
+  }
+});
+
+await assert.rejects(
+  () =>
+    env.run(async (x) => {
+      Context.current().heartbeat(6);
+      await Context.current().sleep(100); // <- sleep is cancellation aware
+    }, 3),
+  (err) => {
+    assert.ok(err instanceof CancelledFailure);
+  }
+);
+```
+
+## Testing Workflows
+
+Workflows can be tested with [`TestWorkflowEnvironment`](https://typescript.temporal.io/api/classes/testing.TestWorkflowEnvironment).
+
+A typical test suite would set up a single instance of the test environment to be reused in all tests (e.g. in a [jest](https://jestjs.io/) `beforeAll` hook).
+
+When creating an environment [`TestWorkflowEnvironment.create`](https://typescript.temporal.io/api/classes/testing.TestWorkflowEnvironment#create) it will automatically start a test server which you can access with [`workflowClient`](https://typescript.temporal.io/api/classes/testing.TestWorkflowEnvironment#workflowclient) and [`nativeConnection`](https://typescript.temporal.io/api/classes/testing.TestWorkflowEnvironment#nativeconnection).
 
-  it('retries one failure', async function () {
-    // Make the first request fail, but subsequent requests succeed
-    let numCalls = 0;
-    sinon.stub(axios, 'get').callsFake(() => {
-      if (numCalls++ === 0) {
-        return Promise.reject(new Error('first error'));
-      }
-      return Promise.resolve({ data: { args: { answer: '88' } } });
-    });
-
-    const result = await client.execute(example);
-    assert.equal(result, 'The answer is 88');
+### Example setup
+
+> NOTE: `beforeAll` and `afterAll` are injected by `jest`.
```

Comment:
If you prefer: `import { beforeAll, afterAll } from '@jest/globals'`

Reply:
🤷